rtrlib/bgpsec/bgpsec.c

```diff
@@ -0,0 +1,928 @@
+
+	/* A temporare spki record */
+	struct spki_record *tmp_key = NULL;
+
+	/* Check, if the parameters are not NULL */
+	if (!data || !sig_segs || !sec_paths || !table)
+		return RTR_BGPSEC_ERROR;
+
+	/* Check if there has been at least one hop */
+	if (as_hops < 1)
+		return RTR_BGPSEC_ERROR;
+
+	/* Check, if the algorithm suite is supported by RTRlib. */
+	if (rtr_bgpsec_has_algorithm_suite(data->alg_suite_id) ==
+			RTR_BGPSEC_ERROR) {
+		return RTR_BGPSEC_UNSUPPORTED_ALGORITHM_SUITE;
+	}
+
+	/* Make sure that all router keys are available. */
+	for (unsigned int i = 0; i < as_hops; i++) {
+		unsigned int router_keys_len = 0;
+		enum spki_rtvals spki_retval = spki_table_search_by_ski(
+							table,
+							sig_segs[i].ski,
+							&tmp_key,
+							&router_keys_len);
+		if (spki_retval == SPKI_ERROR)
+			goto err;
+
+		/* Return an error, if a router key was not found. */
+		if (router_keys_len == 0) {
+			char ski_str[SKI_STR_LEN] = {'\0'};
+
+			ski_to_char(ski_str, sig_segs[i].ski);
+			BGPSEC_DBG(
+				"ERROR: Could not find router key for SKI: %s",
+				ski_str);
+			goto err;
+		}
+		lrtr_free(tmp_key);
+	}
+
+	retval = align_val_byte_sequence(data, sig_segs, sec_paths,
+					 as_hops, &bytes, &bytes_len);
+
+	if (retval == RTR_BGPSEC_ERROR)
+		goto err;
+
+	hash_result = lrtr_malloc(SHA256_DIGEST_LENGTH);
+	if (!hash_result)
+		goto err;
+
+	/*
+	 * offset: the current position from where bytes should
+	 * be processed.
+	 *
+	 * next_offset: adds to offset, after the bytes on the
+	 * current offset have been processed. next_offset is not
+	 * constant and must be calculated each iteration:
+	 *
+	 * signature length +
+	 * SKI size +
+	 * sizeof(var that holds signature length) +
+	 * sizeof(a secure path segment)
+	 *
+	 *
+	 *  offset
+	 * |o----------------------------------------| bytes
+	 *
+	 *		offset+=
+	 *		new_offset
+	 * |++++++++++++o----------------------------| bytes
+	 *
+	 *			    offset+=
+	 *			    new_offset
+	 * |++++++++++++++++++++++++o----------------| bytes
+	 *
+	 *
+	 * A more detailed view can be found at
+	 * https://mailarchive.ietf.org/arch/msg/sidr/8B_e4CNxQCUKeZ_AUzsdnn2f5Mu
+	 **/
+
+	/* Set retval to RTR_BGPSEC_VALID so the for-condition does not
+	 * fail on the first time checking.
+	 */
+	retval = RTR_BGPSEC_VALID;
+
+	for (unsigned int i = 0, offset = 0, next_offset = 0;
+	     offset <= bytes_len && retval == RTR_BGPSEC_VALID;
+	     offset += next_offset, i++) {
+		next_offset =	sig_segs[i].sig_len +
+				SKI_SIZE +
+				sizeof(sig_segs[i].sig_len) +
+				sizeof(struct rtr_secure_path_seg);
+
+		/* Hash the bytes from the offset position out. */
+		retval = hash_byte_sequence(&bytes[offset],
+					    (bytes_len - offset),
+					    data->alg_suite_id,
+					    hash_result);
+
+		if (retval != RTR_BGPSEC_SUCCESS)
+			goto err;
+
+		/* Store all router keys for the given SKI in tmp_key. */
+		unsigned int router_keys_len = 0;
+		enum spki_rtvals spki_retval = spki_table_search_by_ski(
```

Comment:
Why do another lookup?
You could just store them when checking for their presence.

Reply:
When I store the keys during presence check, I would later have to look them up from that storage again, no matter what. I feel like this would substitute one lookup with another. Or am I missing something here?